File: ui/main/src/app/services/push-notification.service.ts
/* Copyright (c) 2022, RTE (http://www.rte-france.com)
 * See AUTHORS.txt
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 * This file is part of the OperatorFabric project.
 */

import {environment} from '@env/environment';
import {HttpClient, HttpErrorResponse} from '@angular/common/http';
import {catchError} from 'rxjs/operators';
import {Injectable} from '@angular/core';
import {ErrorService} from '@ofServices/error-service';
import {Store} from '@ngrx/store';
import {AppState} from '@ofStore/index';
import {OpfabLoggerService} from './logs/opfab-logger.service';

@Injectable({
    providedIn: 'root'
})
export class PushNotificationService extends ErrorService {
    readonly pushNotificationServiceUrl: string;

    /**
     * @constructor
     * @param httpClient - Angular build-in
     */
    constructor(protected store: Store<AppState>, private httpClient: HttpClient, protected loggerService: OpfabLoggerService) {
        super(store, loggerService);
        this.pushNotificationServiceUrl = `${environment.urls.pushNotification}`;
    }


    public sendSubscriptionToServer(subscription: PushSubscription) {

        // Get public key and user auth from the subscription object
        var key = subscription.getKey('p256dh') ? subscription.getKey('p256dh') : '';
        var auth = subscription.getKey('auth') ? subscription.getKey('auth') : '';

        const subscriptionObject = new SubscriptionObject();
        subscriptionObject.endpoint = subscription.endpoint;

        subscriptionObject.key = subscription.getKey('p256dh')? btoa(String.fromCharCode.apply(null, new Uint8Array(subscription.getKey('p256dh')))) : ''
        subscriptionObject.auth = subscription.getKey('auth')? btoa(String.fromCharCode.apply(null, new Uint8Array(subscription.getKey('auth')))) : ''
        console.log("Send subscription to server " + subscriptionObject);
        return this.httpClient
            .post<SubscriptionObject>(`${this.pushNotificationServiceUrl}/subscription`, subscriptionObject)
            .pipe(catchError((error: HttpErrorResponse) => this.handleError(error)));
    }

    public deleteSubscription(subscription: PushSubscription) {
        const encodedEndpoint = this.toBase64(subscription.endpoint);
        const url = `${this.pushNotificationServiceUrl}/subscription/${encodedEndpoint}`;
        return this.httpClient.delete(url).pipe(
            catchError((error: HttpErrorResponse) => this.handleError(error))
        );
    }

    private toBase64(str) {
        return window.btoa(encodeURIComponent(str));
    }

}

class SubscriptionObject {
    public endpoint: string;
    public key: string;
    public auth: string;
}

File: services/cards-consultation/src/main/java/org/opfab/cards/consultation/services/PushNotificationService.java
/* Copyright (c) 2022, RTE (http://www.rte-france.com)
 * See AUTHORS.txt
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 * This file is part of the OperatorFabric project.
 */

package org.opfab.cards.consultation.services;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import nl.martijndwars.webpush.Notification;
import nl.martijndwars.webpush.PushService;
import org.apache.http.HttpResponse;
import org.jose4j.lang.JoseException;
import org.opfab.cards.consultation.model.PushSubscriptionData;
import org.opfab.cards.consultation.repositories.PushSubscriptionRepository;
import org.opfab.springtools.error.model.ApiError;
import org.opfab.springtools.error.model.ApiErrorException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.security.GeneralSecurityException;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.spec.InvalidKeySpecException;
import java.util.Optional;
import java.util.concurrent.ExecutionException;

@Service
@Slf4j
public class PushNotificationService {
    @Autowired
    private PushSubscriptionRepository pushSubscriptionRepository;

    @Autowired
    private ObjectMapper objectMpper;

    @Value("${webPush.privateKey}")
    private String privateKey;

    @Value("${webPush.publicKey}")
    private String publicKey;

    @Value("${webPush.subject}")
    private String subject;

    public void savePushSubscription(PushSubscriptionData subscription) {
        pushSubscriptionRepository.save(subscription);
    }


    public void deleteSubscription(String endpoint, String login) {
        Optional<PushSubscriptionData> sub = findSubscription(endpoint);
        if (sub.isPresent()) pushSubscriptionRepository.delete(sub.get());
        else log.info("PushSubscription not found for user : " + login + " and endpoint:" + endpoint);
    }

    private Optional<PushSubscriptionData> findSubscription(String endpoint) {
        return pushSubscriptionRepository.findById(endpoint);
    }

    public void pushMessageToSubscriptions(String message) {
        pushSubscriptionRepository.findAll().forEach(s -> pushMessageToSubscription(message, s)
        );
    }

    private void pushMessageToSubscription(String message, PushSubscriptionData subscription) {
        Notification notification;
        PushService pushService = new PushService();
        try {
            pushService.setPrivateKey(privateKey);

            pushService.setPublicKey(publicKey);

            pushService.setSubject(subject);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        } catch (NoSuchProviderException e) {
            throw new RuntimeException(e);
        } catch (InvalidKeySpecException e) {
            throw new RuntimeException(e);
        }


        OpfabNotification notify = new OpfabNotification();
        notify.setTitle("OpFab notification");
        notify.setBody(message);
        Payload payload = new Payload(notify);



        // Create a notification with the endpoint, userPublicKey from the subscription and a custom payload
        try {
            String jsonPayload = objectMpper.writeValueAsString(payload);

            //"{\"notification\":{\"title\":\"OpFab notification\", \"body\":\"" + message +" \"}}";
            log.info("Sending notification to " + subscription.getLogin() + " payload = " + jsonPayload );

            notification = new Notification(
                    subscription.getEndpoint(),
                    subscription.getUserPublicKey(),
                    subscription.getAuthAsBytes(),
                    jsonPayload.getBytes()
            );
            HttpResponse res = pushService.send(notification);
            log.info("Notification sent response " + res);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Data
    @AllArgsConstructor
    class Payload {
        private OpfabNotification notification;
    }

    @Data
    class OpfabNotification {
        String title;
        String body;
    }
}
